### ISSUE TITLE
Switch to nginx

### ISSUE THREAD

**Issue by benjaminjkraft** [learning-unlimited/ESP-Website#2044]
We may want to consider switching the servers to use nginx instead of Apache, since these days that's the more standard stack -- it's probably faster and less of a memory hog, and is less painful to configure correctly.  However, that will involve migrating all of our config and scripts.

I think MIT ESP already did this (cc @fbastani)?

**Comment by uakfdotb** [comment]
Yeah, we are running nginx and uwsgi.

nginx handling redirects to SSL:

	server {
		listen 80;
		server_name [...];
		return 301 https://$request_uri;
	}

nginx proxying from SSL to varnish:

	server {
		listen 443 ssl;
		server_name [...];

		# SSL settings from https://wiki.mozilla.org/Security/Server_Side_TLS#Nginx
		# ...

		location / {
			proxy_pass http://localhost:[varnish port];
			proxy_set_header Host $host;
			proxy_set_header X-Real-IP $remote_addr;
			proxy_set_header X-Forwarded-for $remote_addr;
			proxy_read_timeout 15m;
		}
	}

Then, varnish forwards request back to nginx on another port, and nginx proxies to uwsgi:

	server {
		listen 127.0.0.1:[some port];
		server_name [...];

		location /robots.txt {
			return 301 /media/robots.txt;
		}
		location /favicon.ico {
			return 301 /media/favicon.ico;
		}

		# max upload size
		client_max_body_size 64M;

		# Django media
		location /media {
			rewrite ^/media/uploaded/([^/]*)/([^/]*)$ /download/$2;
			alias /path/to/esp/public/media;
		}

		location /static {
			alias /path/to/esp/public/static;
		}

		# uwsgi app
		location / {
			uwsgi_pass unix:///run/uwsgi/app/SITENAME/socket;
			include /etc/nginx/uwsgi_params;
			uwsgi_read_timeout 15m;

			# if HTTPS, ESP-Website needs to know or there'll be redirect loop issues for some program modules
			uwsgi_param HTTP_X_FORWARDED_PROTOCOL https;
		}
	}

And finally our uwsgi configuration is pretty simple (using uwsgi package on Ubuntu, you can just stick it in /etc/uwsgi/apps-enabled/SITENAME.ini):

	[uwsgi]
	wsgi-file = /path/to/esp/esp.wsgi
	master = true
	processes = 12
	vacuum = true
	buffer-size = 16384 # fix issues with size of HTTP headers being too high for some browsers

**Comment by willgearty** [comment]
@milescalabresi maybe useful for AWS?

**Comment by milescalabresi** [comment]
I don't think it helps with AWS migration beyond beyond the benefits Ben mentioned

**Comment by willgearty** [comment]
Sorry, I should clarify. Since we are migrating to AWS, maybe it would be a good time to do this @milescalabresi ?

**Comment by milescalabresi** [comment]
I'm not convinced there's an advantage to doing it now vs in the future. My preference is to finish the AWS migration on Apache then discuss moving to nginx

**Comment by Dhirenderchoudhary** [comment]
@willgearty if issue not resolved i'll work on it

**Comment by Supriyo760** [comment]
@milescalabresi @willgearty 
Before diving into this, could one of the maintainers confirm if the AWS migration discussed previously has been completed? 
If the infrastructure is ready, I'd be happy to start working on standardizing our configurations to migrate from Apache to Nginx. Let me know if this is something the core team wants to prioritize right now!

**Comment by willgearty** [comment]
We are migrated to AWS. I'll defer to @milescalabresi on the possibility of migrating to nginx (and if that would even be possible for someone within server access to work on).
